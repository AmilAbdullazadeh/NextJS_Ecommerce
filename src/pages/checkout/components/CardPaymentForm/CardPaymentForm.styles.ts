import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(({ breakpoints }) => ({
    cardImage: {
        [breakpoints.down('xs')]: {
            display: 'none',
        },
    },
}));

export default useStyles;
