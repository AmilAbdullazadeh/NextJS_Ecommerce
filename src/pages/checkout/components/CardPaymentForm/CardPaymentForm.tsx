import React, { useEffect, useState } from 'react';
import Cards from 'react-credit-cards';
import 'react-credit-cards/es/styles-compiled.css';
import { TextField } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';

import {
    useDebounce,
    useAppDispatch,
    useAppSelector as useSelect,
} from '@hooks';
import {
    updateCardCvcAction,
    updateCardExpiryAction,
    updateCardNameAction,
    updateCardNumberAction,
} from '@pages/checkout/store/actions';
import {
    cardCvcSelector,
    cardExpirySelector,
    cardNameSelector,
    cardNumberSelector,
    isValidCardCvcSelector,
    isValidCardExpirySelector,
    isValidCardNameSelector,
    isValidCardNumberSelector,
} from '@pages/checkout/store/selectors';

import useStyles from './CardPaymentForm.styles';

const CardPaymentForm: React.FC = () => {
    const inputDebounce = 250;
    const classes = useStyles();
    const dispatch = useAppDispatch();

    const cardNumber = useSelect(cardNumberSelector);
    const cardName = useSelect(cardNameSelector);
    const cardExpiry = useSelect(cardExpirySelector);
    const cardCvc = useSelect(cardCvcSelector);

    const isValidCardNumber = useSelect(isValidCardNumberSelector);
    const [cardNumberError, setCardNumberError] = useState(false);

    const isValidCardName = useSelect(isValidCardNameSelector);
    const [cardNameError, setCardNameError] = useState(false);

    const isValidCardExpiry = useSelect(isValidCardExpirySelector);
    const [cardExpiryError, setCardExpiryError] = useState(false);

    const isValidCardCvc = useSelect(isValidCardCvcSelector);
    const [cardCvcError, setCardCvcError] = useState(false);

    const [cardInfo, setCardInfo] = useState({
        number: cardNumber,
        name: cardName,
        expiry: cardExpiry,
        cvc: cardCvc,
    });

    const debouncedCardNumber = useDebounce(cardInfo.number, inputDebounce);
    const debouncedCardName = useDebounce(cardInfo.name, inputDebounce);
    const debouncedCardExpiry = useDebounce(cardInfo.expiry, inputDebounce);
    const debouncedCardCvc = useDebounce(cardInfo.cvc, inputDebounce);

    useEffect(() => {
        dispatch(updateCardNumberAction({ number: debouncedCardNumber }));
    }, [debouncedCardNumber]);

    useEffect(() => {
        dispatch(updateCardNameAction({ name: debouncedCardName }));
    }, [debouncedCardName]);

    useEffect(() => {
        dispatch(updateCardExpiryAction({ expiry: debouncedCardExpiry }));
    }, [debouncedCardExpiry]);

    useEffect(() => {
        dispatch(updateCardCvcAction({ cvc: debouncedCardCvc }));
    }, [debouncedCardCvc]);

    const handleNumberChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setCardInfo(prevInfo => ({
            ...prevInfo,
            number: e.target.value,
        }));

    const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setCardInfo(prevInfo => ({
            ...prevInfo,
            name: e.target.value,
        }));

    const handleExpiryChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setCardInfo(prevInfo => ({
            ...prevInfo,
            expiry: e.target.value,
        }));

    const handleCvcChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setCardInfo(prevInfo => ({
            ...prevInfo,
            cvc: e.target.value,
        }));

    const textFieldProps = [
        {
            label: 'card number',
            onChange: handleNumberChange,
            value: cardInfo.number,
            onBlur: () => setCardNumberError(!isValidCardNumber),
            error: cardNumberError,
            helperText: cardNumberError
                ? 'field should contain 16 numbers'
                : '',
        },
        {
            label: 'card name',
            onChange: handleNameChange,
            value: cardInfo.name,
            onBlur: () => setCardNameError(!isValidCardName),
            error: cardNameError,
            helperText: cardNameError
                ? 'field should contain a valid name'
                : '',
        },
        {
            label: 'expiry',
            onChange: handleExpiryChange,
            value: cardInfo.expiry,
            onBlur: () => setCardExpiryError(!isValidCardExpiry),
            error: cardExpiryError,
            helperText: cardExpiryError
                ? 'field should contain a valid expiry date'
                : '',
        },
        {
            label: 'cvc',
            onChange: handleCvcChange,
            value: cardInfo.cvc,
            type: 'password',
            onBlur: () => setCardCvcError(!isValidCardCvc),
            error: cardCvcError,
            helperText: cardCvcError ? 'field should contain a valid cvc' : '',
        },
    ];

    return (
        <div id="payment-info-form" data-testid="payment-info-form">
            <Grid
                container
                direction="column"
                justifyContent="center"
                alignItems="center"
                spacing={2}
            >
                <Grid item xs={12}>
                    <div className={classes.cardImage}>
                        <Cards
                            cvc={cardInfo.cvc}
                            expiry={cardInfo.expiry}
                            name={cardInfo.name}
                            number={cardInfo.number}
                        />
                    </div>
                </Grid>
                {textFieldProps.map(
                    ({ label, onChange, error, value, ...rest }) => (
                        <Grid item xs={12} key={label}>
                            <TextField
                                label={label}
                                variant="outlined"
                                color="primary"
                                error={error}
                                onChange={onChange}
                                inputProps={{
                                    'aria-label': label,
                                    'data-testid': label,
                                }}
                                value={value}
                                {...rest}
                            />
                        </Grid>
                    )
                )}
            </Grid>
        </div>
    );
};

export default CardPaymentForm;
