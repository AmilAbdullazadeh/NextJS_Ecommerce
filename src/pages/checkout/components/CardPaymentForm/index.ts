export { default as CardPaymentForm } from './CardPaymentForm';
