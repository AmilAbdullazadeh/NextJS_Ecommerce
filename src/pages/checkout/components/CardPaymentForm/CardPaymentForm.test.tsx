import React from 'react';
import faker from 'faker';

import { renderWithProviders, fireEvent, waitFor } from '@testUtils';

import CardPaymentForm from './CardPaymentForm';

describe('CardPaymentForm', () => {
    it('renders correctly', () => {
        const {
            container: { firstChild },
        } = renderWithProviders(<CardPaymentForm />);

        expect(firstChild).toMatchSnapshot();
    });
    it('updates card number', () => {
        const { getByRole } = renderWithProviders(<CardPaymentForm />);

        const cardNumberInput = getByRole('textbox', {
            name: /card number/i,
        });

        const input = '1234123412341234';

        fireEvent.change(cardNumberInput, {
            target: { value: input },
        });

        expect(cardNumberInput).toHaveAttribute('value', input);
    });
    it('updates card name', () => {
        const { getByRole } = renderWithProviders(<CardPaymentForm />);

        const cardNameInput = getByRole('textbox', {
            name: /card name/i,
        });

        const input = `${faker.name.firstName()} ${faker.name.lastName()}`;

        fireEvent.change(cardNameInput, {
            target: { value: input },
        });

        expect(cardNameInput).toHaveAttribute('value', input);
    });

    it('updates card expiry', () => {
        const { getByRole } = renderWithProviders(<CardPaymentForm />);

        const cardExpiryInput = getByRole('textbox', { name: /expiry/i });

        const input = '1299';

        fireEvent.change(cardExpiryInput, {
            target: { value: input },
        });

        expect(cardExpiryInput).toHaveAttribute('value', input);
    });
    it('updates card cvc', () => {
        const { getByTestId } = renderWithProviders(<CardPaymentForm />);

        const cvcInput = getByTestId('cvc');

        const input = '123';

        fireEvent.change(cvcInput, {
            target: { value: input },
        });

        expect(cvcInput).toHaveAttribute('value', input);
    });

    it('displays card number error', async () => {
        const { getByRole, getByText } = renderWithProviders(
            <CardPaymentForm />
        );

        const cardNumberInput = getByRole('textbox', {
            name: /card number/i,
        });

        cardNumberInput.focus();

        cardNumberInput.blur();

        await waitFor(() =>
            expect(getByText('field should contain 16 numbers')).toBeVisible()
        );
    });
    it('displays card name error', async () => {
        const { getByRole, getByText } = renderWithProviders(
            <CardPaymentForm />
        );

        const cardNameInput = getByRole('textbox', {
            name: /card name/i,
        });

        cardNameInput.focus();

        cardNameInput.blur();

        await waitFor(() =>
            expect(getByText('field should contain a valid name')).toBeVisible()
        );
    });
    it('displays card expiry error', async () => {
        const { getByRole, getByText } = renderWithProviders(
            <CardPaymentForm />
        );

        const cardExpiryInput = getByRole('textbox', { name: /expiry/i });

        cardExpiryInput.focus();

        cardExpiryInput.blur();

        await waitFor(() =>
            expect(
                getByText('field should contain a valid expiry date')
            ).toBeVisible()
        );
    });

    it('displays card cvc error', async () => {
        const { getByTestId, getByText } = renderWithProviders(
            <CardPaymentForm />
        );

        const cvcInput = getByTestId('cvc');

        cvcInput.focus();

        cvcInput.blur();

        await waitFor(() =>
            expect(getByText('field should contain a valid cvc')).toBeVisible()
        );
    });
});
