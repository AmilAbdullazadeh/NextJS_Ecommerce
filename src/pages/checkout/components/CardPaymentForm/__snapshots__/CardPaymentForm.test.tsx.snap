// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPaymentForm renders correctly 1`] = `
<div
  data-testid="payment-info-form"
  id="payment-info-form"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-cardImage-1"
      >
        <div
          class="rccs"
        >
          <div
            class="rccs__card rccs__card--unknown"
          >
            <div
              class="rccs__card--front"
            >
              <div
                class="rccs__card__background"
              />
              <div
                class="rccs__issuer"
              />
              <div
                class="rccs__cvc__front"
              />
              <div
                class="rccs__number"
              >
                •••• •••• •••• ••••
              </div>
              <div
                class="rccs__name"
              >
                YOUR NAME HERE
              </div>
              <div
                class="rccs__expiry"
              >
                <div
                  class="rccs__expiry__valid"
                >
                  valid thru
                </div>
                <div
                  class="rccs__expiry__value"
                >
                  ••/••
                </div>
              </div>
              <div
                class="rccs__chip"
              />
            </div>
            <div
              class="rccs__card--back"
            >
              <div
                class="rccs__card__background"
              />
              <div
                class="rccs__stripe"
              />
              <div
                class="rccs__signature"
              />
              <div
                class="rccs__cvc"
              />
              <div
                class="rccs__issuer"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          card number
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="card number"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="card number"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                card number
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          card name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="card name"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="card name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                card name
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          expiry
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="expiry"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="expiry"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                expiry
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          cvc
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="cvc"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="cvc"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                cvc
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
  </div>
</div>
`;
